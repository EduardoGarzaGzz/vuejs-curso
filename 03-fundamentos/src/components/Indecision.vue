<template>
  <img v-if="img" :src="img" alt="bg">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input v-model="question" placeholder="Hazme una pregunta?" type="text">
    <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>

    <div>
      <h2>{{ question }}</h2>
      <h1 v-if="isValidQuestion">{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question       : null,
      answer         : null,
      img            : null,
      isValidQuestion: false
    }
  },
  methods: {
    async getAnswer() {
      this.answer = 'Pensando...';

      const { answer, image } = await fetch( 'https://yesno.wtf/api' ).then( r => r.json() );

      this.img = image;
      switch ( answer ) {
        case 'yes':
          this.answer = 'SI!';
          break;
        case 'no':
          this.answer = 'NO!';
          break;
        case 'maybe':
          this.answer = 'TAL VES!';
          break;
      }
    }
  },
  watch  : {
    question( value, oldValue ) {
      this.isValidQuestion = false;

      if ( !value.includes( '?' ) ) return;

      this.getAnswer();
      this.isValidQuestion = true;
    }
  }
}
</script>

<style>

    img, .bg-dark {
      height: 100vh;
      left: 0px;
      max-height: 100%;
      max-width: 100%;
      position: fixed;
      top: 0px;
      width: 100vw;
    }

    .bg-dark {
      background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
      position: relative;
      z-index: 99;
    }

    input {
      width: 250px;
      padding: 10px 15px;
      border-radius: 5px;
      border: none;
    }

    input:focus {
      outline: none;
    }

    p {
      color: white;
      font-size: 20px;
      margin-top: 0px;
    }

    h1, h2 {
      color: white;
    }

    h2 {
      margin-top: 150px;
    }

</style>
