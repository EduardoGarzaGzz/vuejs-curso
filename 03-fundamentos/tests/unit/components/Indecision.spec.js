import Indecision       from "@/components/Indecision";
import { shallowMount } from "@vue/test-utils";

describe( 'Indecision Component', () => {
	global.fetch = jest.fn( () => Promise.resolve( {
		json: () => Promise.resolve( {
			"answer": "yes",
			"forced": false,
			"image" : "https://yesno.wtf/assets/yes/5-64c2804cc48057b94fd0b3eaf323d92c.gif"
		} )
	} ) );
	
	let wrapper = null;
	
	beforeEach( () => {
		wrapper = shallowMount( Indecision );
		jest.clearAllMocks();
	} );
	
	it( 'debe de hacer march con el snapshot', () => {
		expect( wrapper.html() ).toMatchSnapshot();
	} );
	
	it( 'escribir en el input no debe de disparar nada (console.log)', async () => {
		const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' );
		const clgSpy       = jest.spyOn( console, 'log' );
		const inp          = wrapper.find( 'input' );
		
		await inp.setValue( 'Hola mundo' );
		
		expect( clgSpy ).toHaveBeenCalledTimes( 1 );
		expect( getAnswerSpy ).toHaveBeenCalledTimes( 0 );
	} );
	
	it( 'escribir el symbols de "?" debe de disparar e; fetch', async () => {
		const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' );
		const inp          = wrapper.find( 'input' );
		
		await inp.setValue( 'Hola mundo?' );
		
		expect( getAnswerSpy ).toHaveBeenCalledTimes( 1 );
	} );
	
	it( 'pruebas en getAnswer', async () => {
		await wrapper.vm.getAnswer();
		
		const img = wrapper.find( 'img' );
		
		expect( img.exists() ).toBeTruthy();
		expect( wrapper.vm.img ).toBe( 'https://yesno.wtf/assets/yes/5-64c2804cc48057b94fd0b3eaf323d92c.gif' );
		expect( wrapper.vm.answer ).toBe( 'SI!' );
	} );
	
	it( 'pruebas en getAnswer - Fallo en el API', async () => {
		fetch.mockImplementationOnce( () => Promise.reject( 'API is down' ) );
		await wrapper.vm.getAnswer();
		
		const img = wrapper.find( 'img' );
		
		expect( img.exists() ).toBeFalsy();
		expect( wrapper.vm.answer ).toBe( 'No se pudo cargar del API' );
	} );
} );
