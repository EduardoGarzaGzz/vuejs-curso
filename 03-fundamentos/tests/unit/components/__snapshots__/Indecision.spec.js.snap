// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe de hacer march con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input placeholder="Hazme una pregunta?" type="text">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <div>
    <h2></h2>
    <!--v-if-->
  </div>
</div>
`;
